### PR TITLE
Allow passing the current value to config flow input fields

### LINKED ISSUE #5337: Allowing optional values in options flow to be removed
## The request

When using optional values in an options flow, and giving them default values, there's no real way to remove them - if the user deletes the value, the default one (which would usually be populated with the previous value) will be used.

The Monoprice options flow is an example of this at work - in order to remove a value you have to replace it with whitespace instead of deleting, thus working around the default value, and the integration will take then care of ignoring the whitespace.

My proposal is to have some mechanism to change that behavior. The two options I can think of:
1. As proposed by @balloob in https://github.com/home-assistant/core/pull/33156#issuecomment-603378610, have some `allow_empty` indication
2. Separate between the default value, and the pre-populated value, so you can pre-populate the existing options without setting default.

Neither is supported by Voluptuous, so we will need some other mechanism to pass this information along.

## The alternatives

I have two alternatives mentioned above, and would love to hear more.

### PR DESCRIPTION
**Pull request by OnFreund**
## Proposed change

This PR creates a mechanism for the backend to pass the current value of config flow fields. The existing way of doing this is to set the `default` parameter in the schema, but that has the adverse effect of not being able to remove an optional value, since it will be populated with the default if left empty.
After this PR is merged, the backend can pass the current value without setting a default. An example:

```python
vol.Schema({
  vol.Optional("port", description={"suggested_value": 1234}): int
})
```
If the current value is not present, the default will still be used, so existing integrations would not be broken as a result of this change.

You can also mix and match the current value and the default - the dialog will show the current value, but if left empty, the default will be used.
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #5337 
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-form/ha-form.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export interface HaFormBaseSchema {
   default?: HaFormData;
   required?: boolean;
   optional?: boolean;
-  description?: { suffix?: string };
+  description?: { suffix?: string; current_value?: any };
```
Comment (bramkragten) [inline src/components/ha-form/ha-form.ts:33]:
```suggestion
  description?: { suffix?: string; current_value?: HaFormData };
```

File: src/dialogs/config-flow/step-flow-form.ts [hunk 2]
```diff
@@ -122,7 +122,9 @@ class StepFlowForm extends LitElement {
 
     const data = {};
     this.step.data_schema.forEach((field) => {
-      if ("default" in field) {
+      if (field?.description?.current_value) {
```
Comment (bramkragten) [inline src/dialogs/config-flow/step-flow-form.ts:125]:
```suggestion
      if (field.description?.current_value) {
```

**Comment by bramkragten** [comment]
This should also be added to all the `ha-form-*` elements.

**Comment by OnFreund** [comment]
Thanks @bramkragten.

> This should also be added to all the ha-form-* elements.

Not sure what you mean. The `ha-form-*` elements see the current value already pre-filled, so I'm not sure what change is required if at all.

**Comment by bramkragten** [comment]
I have to check how this all works, but you now do it in only one place. The device automations might also need this? And at least `ha-form-integer` now uses `default` so should be updated?

**Comment by bramkragten** [comment]
Looks like device automations now don't display the default at all, only the set value if it has it. I personally would like to remove the logic of setting the data based on `default`/`current_value` in `step-flow-form` and move it to `ha-form`.

But that might be out of scope for this PR.

**Comment by OnFreund** [comment]
1. I renamed `current_value` to `suggested_value` - I think it reflects more uses cases (e.g. the one in https://github.com/home-assistant/core/pull/33621.

2. Device Automations - I don't have much experience there, so I don't really know, but I believe that enabling this for config and options flows is already great value on its own. As for moving it into `ha-form`, it sounds right, but TBH I don't believe I understand this code well enough to pull this off. I agree with you it could out of scope here.

3. I'm not sure why `ha-form-integer` has its own handling of `default`, but it only kicks in when there's no data, which will not happen if there's a `default` or `suggested_value` in a config/options flow. I assumed that this might be used by other clients (e.g. device automations that you brought up), so I preferred not to touch it and break something else.

**Comment by bramkragten** [comment]
It is because `ha-form-integer` already supports optional based on the optional flag

**Comment by bramkragten** [comment]
This does not depend on a backend change right?

**Review by bramkragten — APPROVED** [review]

**Comment by OnFreund** [comment]
No backend change required - this is completely backwards compatible.
However, integrations that currently use `default` for optional values, can migrate to `suggested_value` if it makes sense for them.

**Comment by OnFreund** [comment]
@bramkragten I think there might be a bug here that I missed because my backend was covering up for it - if the user changes nothing in the form and submits, the suggested value isn't used. My best suggestion on how to fix this is to end `_stepDataProcessed()` with:
```javascript
this._stepData = data;
return data;
```
But I'm assuming there's a reason it's not there? Maybe we can set it only in case a suggested value was picked up?

**Comment by OnFreund** [comment]
Something like:
```javascript
private get _stepDataProcessed() {
    if (this._stepData !== undefined) {
      return this._stepData;
    }

    const data = {};
    suggestion_used = false;
    this.step.data_schema.forEach((field) => {
      if (field.description?.suggested_value) {
        data[field.name] = field.description.suggested_value;
        suggestion_used = true;
      } else if ("default" in field) {
        data[field.name] = field.default;
      }
    });

    if (suggestion_used) {
      this._stepData = data;
    }
    return data;
  }

```

**Comment by bramkragten** [comment]
I think we can just set it, it only means the frontend is doing the backends work of setting defaults.

**Comment by OnFreund** [comment]
OK, new PR here: https://github.com/home-assistant/frontend/pull/5688

**Pull request merged**